core/services/capreg/launcher.go

```diff
@@ -0,0 +1,174 @@
+			removed, newOrUpdated := capabilityDiff(l.myDONs[don.ID].CapabilityConfigurations, don.CapabilityConfigurations)
+			for _, capabilityID := range newOrUpdated {
+				capSvc, ok2 := l.capabilitySvcs[capabilityID]
+				if !ok2 {
+					errs = multierr.Append(errs, fmt.Errorf("update: capability service not found for capability ID %s", capabilityID))
+					continue
+				}
+
+				if err := capSvc.Update(ctx, don); err != nil {
+					errs = multierr.Append(errs,
+						fmt.Errorf("failed to update capability service %s with new DON configuration: %w, DON: %+v",
+							capabilityID, err, don))
+				}
+			}
+			for _, capabilityID := range removed {
+				capSvc, ok2 := l.capabilitySvcs[capabilityID]
+				if !ok2 {
+					errs = multierr.Append(errs, fmt.Errorf("update: capability service not found for capability ID %s", capabilityID))
+					continue
+				}
+
+				if err := capSvc.Stop(ctx, don); err != nil {
+					errs = multierr.Append(errs,
+						fmt.Errorf("failed to stop capability service %s with updated DON configuration: %w, DON: %+v",
+							capabilityID, err, don))
+				}
+			}
+
+			l.myDONs[don.ID] = don
+		} else {
+			// New DON, start capability services specified by it.
+			l.myDONs[don.ID] = don
+			for _, cc := range don.CapabilityConfigurations {
+				capSvc, ok2 := l.capabilitySvcs[cc.CapabilityID]
+				if !ok2 {
+					// This shouldn't happen - nodes should only be added to a DON if they support the capability
+					// of that DON.
+					errs = multierr.Append(errs, fmt.Errorf("start: capability service not found for capability ID %s", cc.CapabilityID))
+					continue
+				}
+
+				if err := capSvc.Start(ctx, don); err != nil {
+					errs = multierr.Append(errs,
+						fmt.Errorf("failed to start capability service %s with new DON configuration: %w, DON: %+v",
+							cc.CapabilityID, err, don))
+				}
+			}
+		}
+	}
+	return errs
+}
+
+func (l *capabilityLauncher) handleDeletedDONs(ctx context.Context, relevantDONs map[uint32]DON, errs error) error {
+	for _, don := range l.myDONs {
+		_, found := relevantDONs[don.ID]
+		if !found {
+			// DON present in our state and not onchain state == DON has been deleted from the onchain state.
+			// Spin down capability services for this particular DON.
+			for _, cc := range don.CapabilityConfigurations {
+				capSvc, ok := l.capabilitySvcs[cc.CapabilityID]
+				if !ok {
+					errs = multierr.Append(errs, fmt.Errorf("delete: capability service not found for capability ID %s", cc.CapabilityID))
+					continue
+				}
+
+				if err := capSvc.Stop(ctx, don); err != nil {
+					errs = multierr.Append(errs,
+						fmt.Errorf("failed to stop capability service %s with deleted DON configuration: %w, DON: %+v",
+							cc.CapabilityID, err, don))
+				}
+			}
+			delete(l.myDONs, don.ID)
+		}
+	}
+	return errs
+}
+
+// RegisterCapabilityService registers the provided capability service with the launcher.
+func (l *capabilityLauncher) RegisterCapabilityService(c CapabilityService) error {
```

Comment:
Hmmmmmmmm... it's confusing to register a service I think, since that implies that the service already exists, but the point of this is to launch a service. Should we instead be registering some kind of factory function that returns a CapabilityService?

Reply:
Yeah, good point probably better named to be a factory.